src/main/java/de/tum/in/www1/artemis/service/TutorialGroupService.java

```diff
@@ -0,0 +1,114 @@
+            return; // Registration already exists, nothing to do.
+        }
+        TutorialGroupRegistration newRegistration = new TutorialGroupRegistration(student, tutorialGroup, TutorialGroupRegistrationType.INSTRUCTOR_REGISTRATION);
+        tutorialGroupRegistrationRepository.save(newRegistration);
+    }
+
+    private void registerMultipleStudentsToTutorialGroup(Set<User> students, TutorialGroup tutorialGroup) {
+        Set<User> registeredStudents = tutorialGroupRegistrationRepository.findAllByTutorialGroup(tutorialGroup).stream().map(TutorialGroupRegistration::getStudent)
+                .collect(Collectors.toSet());
+        Set<User> studentsToRegister = students.stream().filter(student -> !registeredStudents.contains(student)).collect(Collectors.toSet());
+        Set<TutorialGroupRegistration> newRegistrations = studentsToRegister.stream()
+                .map(student -> new TutorialGroupRegistration(student, tutorialGroup, TutorialGroupRegistrationType.INSTRUCTOR_REGISTRATION)).collect(Collectors.toSet());
+        tutorialGroupRegistrationRepository.saveAll(newRegistrations);
+    }
+
+    /**
+     * Register multiple students to a tutorial group.
+     *
+     * @param tutorialGroup the tutorial group to register the students for
+     * @param studentDTOs   The students to register.
+     * @return The students that could not be found and thus not registered.
+     */
+    public Set<StudentDTO> registerMultipleStudents(TutorialGroup tutorialGroup, Set<StudentDTO> studentDTOs) {
+        Set<User> foundStudents = new HashSet();
+        Set<StudentDTO> notFoundStudentDTOs = new HashSet<>();
+        for (var studentDto : studentDTOs) {
+            var studentOptional = findStudent(studentDto, tutorialGroup.getCourse().getStudentGroupName());
+            if (studentOptional.isEmpty()) {
+                notFoundStudentDTOs.add(studentDto);
+            }
+            else {
+                foundStudents.add(studentOptional.get());
+            }
+        }
+        registerMultipleStudentsToTutorialGroup(foundStudents, tutorialGroup);
+        return notFoundStudentDTOs;
+    }
+
+    private Optional<User> findStudent(StudentDTO studentDto, String studentCourseGroupName) {
+        var registrationNumber = studentDto.getRegistrationNumber();
+        var login = studentDto.getLogin();
+
+        // try to find the user by login
+        var studentByRegistrationNumber = userRepository.findUserWithGroupsAndAuthoritiesByRegistrationNumber(registrationNumber);
+        if (studentByRegistrationNumber.isPresent()) {
+            var student = studentByRegistrationNumber.get();
+            if (student.getGroups().contains(studentCourseGroupName)) {
+                return studentByRegistrationNumber;
+            }
+        }
+        // try to find the student by login
+        var studentByLogin = userRepository.findUserWithGroupsAndAuthoritiesByLogin(login);
+        if (studentByLogin.isPresent()) {
```

Comment:
Optional: This if statement is a duplicate of the one above. You could combine them.

Reply:
Good idea! I simplified the code logic in: 1ec292a7ac629d1af56794ca12d4627f44a25b27